src/java/org/apache/cassandra/db/compaction/CompactionManager.java

```diff
@@ -845,6 +956,23 @@ public static boolean needsCleanup(SSTableReader sstable, Collection<Range<Token
         return false;
     }
 
+    private boolean checksIfCleanupNeededOne(LifecycleTransaction txn, Collection<Range<Token>> ranges, boolean hasIndexes)
+    {
+        SSTableReader sstable = txn.onlyOne();
+        if (!hasIndexes && !new Bounds<>(sstable.first.getToken(), sstable.last.getToken()).intersects(ranges))
+        {
+            txn.obsoleteOriginals();
+            txn.finish();
+            return false;
+        }
+        if (!needsCleanup(sstable, ranges))
+        {
```

Comment:
This might be a dumb question but I don't fully understand why we don't (need to) close the transaction at this point. From my current understanding of things, it seems that the cleanupOne function does not do it as well.

Reply:
So, I took a further peek and I only think you need to run `txn.finish()` after you run `txn.obsoleteOriginals()`, because you are actually changing the table in that transaction.

For all other cases, the transaction will ultimately get closed in the finally block of the `parallelSstableCallable`:

```
        finally
        {
            Throwable fail = Throwables.close(null, transactions);
            if (fail != null)
                logger.error("Failed to cleanup lifecycle transactions {}", fail);
        }

Reply:
By the way I found this relatively easily by scanning the PR that added Lifecycle transactions (from git blame): https://github.com/apache/cassandra/commit/e5a76bdb5fc04ffa16b8becaa7877186226c3b32